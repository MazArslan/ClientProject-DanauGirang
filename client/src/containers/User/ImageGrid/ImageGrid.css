body {
	text-align: center;
}

.title {
	margin-top: 24px;
	margin-bottom: 2px;
}

.paginationWrapper {
	overflow-x: auto;
}

.image-grid {
	z-index: 25;
	position: sticky;
}

.pagination {
	justify-content: center;
	margin: 30px;
}

.modal-dialog {
	position: relative;
	width: auto;
	pointer-events: none;
	max-width: 500px;
	margin: 1.75rem auto;
}

.card-wrapper {
	margin-top: 10px;
	margin-bottom: 5px;
}

.card-button {
	border: none;
	background-color: transparent;
}

.card-button:focus {
	outline-color: rgb(44, 161, 15);
	outline-style: solid;
}
