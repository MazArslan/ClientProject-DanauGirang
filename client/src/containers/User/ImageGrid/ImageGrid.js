import React, { useState, useEffect, useCallback } from "react";
import { useParams } from "react-router-dom";
import {
  Pagination,
  PaginationItem,
  PaginationLink,
  Container,
  Row,
  Col,
  Button
} from "reactstrap";
import { faFilter } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import useHttp from "../../../hooks/http";

import classes from "./ImageGrid.css";
import { updateObject } from "../../../shared/utility";
import ImageThumbnail from "../../../components/ImageGrid/ImageThumbnail/imageThumbnail";
import Spinner from "../../../components/UI/Spinner/Spinner";
import PageCount from "../../../components/ImageGrid/PageCount/pageCount";
import Filters from "../../../components/ImageGrid/Filters/Filters";
import ImageCardModal from "../ImageCardModal/ImageCardModal";

/**
 * ImageGrid component
 * Displays a paginated view of all of the images for a particular project
 */
const ImageGrid = () => {
  const [captureEvents, setCaptureEvents] = useState([]);
  const [captureEventsId, setCaptureEventsId] = useState();
  const [currentPage, setCurrentPage] = useState(1);
  const [totalPages, setTotalPages] = useState(0);
  const [totalCaptureEvents, setTotalCaptureEvents] = useState(0);
  const [project, setProject] = useState();
  // eslint-disable-next-line no-unused-vars
  const [pageSize, setPageSize] = useState(6);

  const [showFilters, setShowFilters] = useState(false);
  const [filterChoice, setFilterChoice] = useState({
    createdOnDate: "",
    completionStatus: "empty"
  });
  const [filter, setFilter] = useState({
    createdOnDate: "",
    completionStatus: "empty"
  });
  const [selectedCard, setSelectedCard] = useState({ cardId: "" });
  const [showCard, setShowCard] = useState(false);
  const [modal, setModal] = useState(true);
  const { isLoading, error, data, sendRequest } = useHttp();

  const { projectId } = useParams();

  useEffect(() => {
    // eslint-disable-next-line react/prop-types
    const { createdOnDate, completionStatus } = filter;
    const query = `?pageNo=${currentPage}&projectId=${projectId}&completionStatus=${completionStatus}&createdOnDate=${createdOnDate}&pageSize=${pageSize}`;
    sendRequest(
      `http://localhost:3000/api/captureEvent/imageGrid${query}`,
      "GET"
    );
  }, [currentPage, filter, sendRequest, projectId, pageSize]);

  // Assigns the card an id used for ImageCard pagination
  const assignCardId = useCallback(
    captureEventArray => {
      let count = (currentPage - 1) * pageSize;
      captureEventArray.forEach(event => {
        // eslint-disable-next-line no-param-reassign
        event.cardId = count + 1;
        count += 1;
      });
      return captureEventArray;
    },
    [currentPage, pageSize]
  );

  useEffect(() => {
    if (data !== null) {
      setProject(data.project);
      setTotalPages(data.totalPages);
      setTotalCaptureEvents(data.totalCaptureEvents);
      const captureEventThumbnails = assignCardId(data.captureEvents);
      setCaptureEvents(captureEventThumbnails);
      const captureEventsIdArray = [];
      // eslint-disable-next-line array-callback-return
      data.captureEvents.map(captureEvent => {
        captureEventsIdArray.push(captureEvent._id);
      });
      setCaptureEventsId(captureEventsIdArray);
    }
  }, [data, assignCardId, captureEvents]);

  // Pagination bar click handlers
  const handleNumberClick = useCallback((event, index) => {
    event.preventDefault();
    setCurrentPage(index + 1);
  }, []);
  const handleArrowClick = useCallback((event, index) => {
    event.preventDefault();
    setCurrentPage(index);
  }, []);

  const handleThumbnailClick = useCallback(
    (event, cardId) => {
      event.preventDefault();
      const updatedSelectedCard = updateObject(selectedCard, {
        cardId
      });
      setSelectedCard(updatedSelectedCard);
      setModal(true);
      setShowCard(true);
    },
    [selectedCard]
  );

  const toggleModal = useCallback(
    e => {
      e.preventDefault();
      setModal(!modal);
    },
    [modal]
  );

  let imagesGrid = <Spinner />;
  if (!isLoading && captureEvents) {
    imagesGrid = (
      <React.Fragment>
        {captureEvents.map(captureEvent => {
          return (
            <Col xs="6" sm="4" key={captureEvent._id}>
              <div className="card-wrapper">
                <button
                  className="card-button"
                  onClick={event =>
                    handleThumbnailClick(event, captureEvent.cardId)
                  }
                >
                  <ImageThumbnail
                    id={captureEvent._id}
                    image={captureEvent.thumbnailImage}
                    cardId={captureEvent.cardId}
                    dateTime={captureEvent.createdOnDateTime}
                    isComplete={captureEvent.isComplete}
                  />
                </button>
              </div>
            </Col>
          );
        })}
      </React.Fragment>
    );
  }
  let paginationBar = error ? <p>Images cannot be loaded</p> : <Spinner />;
  if (captureEvents) {
    paginationBar = (
      <div className={classes.paginationWrapper}>
        <Pagination aria-label="Page navigation">
          {/* Pagination bar left arrow */}
          <PaginationItem disabled={currentPage <= 1} className="navigate-back">
            <PaginationLink
              onClick={event => handleArrowClick(event, currentPage - 1)}
              previous
              href="#"
            />
          </PaginationItem>
          {/* Pagination bar numbers */}
          {[...Array(totalPages)].map((page, i) => (
            <PaginationItem active={i + 1 === currentPage} key={i + 1}>
              <PaginationLink
                onClick={event => handleNumberClick(event, i)}
                href="#"
              >
                {i + 1}
              </PaginationLink>
            </PaginationItem>
          ))}
          {/* Pagination bar right arrow */}
          <PaginationItem
            disabled={currentPage >= totalPages - 1}
            className="navigate-forward"
          >
            <PaginationLink
              onClick={event => handleArrowClick(event, currentPage + 1)}
              next
              href="#"
            />
          </PaginationItem>
        </Pagination>
      </div>
    );
  }

  let card = null;
  if (showCard && project) {
    card = (
      <ImageCardModal
        cardId={selectedCard.cardId}
        projectId={projectId}
        totalCaptureEvents={totalCaptureEvents}
        pageNumber={currentPage}
        tags={project.tags}
        captureEventsId={captureEventsId}
        pageSize={pageSize}
        isOpen={modal}
        toggle={toggleModal}
      />
    );
  }

  // Functions for the filter state handling logic
  const handleFilterChoiceChange = useCallback(
    (e, type) => {
      const updatedFilter = updateObject(filterChoice, {
        [type]: e.target.value
      });
      setFilterChoice(updatedFilter);
    },
    [filterChoice]
  );

  const handleFiltering = useCallback(
    e => {
      e.preventDefault();
      const updatedFilter = updateObject(filter, {
        createdOnDate: filterChoice.createdOnDate,
        completionStatus: filterChoice.completionStatus
      });
      setFilter(updatedFilter);
    },
    [filter, filterChoice]
  );

  const handleClearFilters = useCallback(
    e => {
      e.preventDefault();
      const updatedFilter = updateObject(filterChoice, {
        createdOnDate: "",
        completionStatus: "empty"
      });
      setFilterChoice(updatedFilter);
      setFilter(updatedFilter);
    },
    [filterChoice]
  );

  const handleShowFilters = useCallback(
    e => {
      e.preventDefault();
      setShowFilters(!showFilters);
    },
    [showFilters]
  );

  let filterComponent = null;
  if (showFilters) {
    filterComponent = (
      <Filters
        filterChoice={filterChoice}
        dateChange={e => handleFilterChoiceChange(e, "createdOnDate")}
        statusChange={e => handleFilterChoiceChange(e, "completionStatus")}
        applyFilters={e => handleFiltering(e)}
        clearFilters={e => handleClearFilters(e)}
      />
    );
  }

  let header = null;
  if (project) {
    header = (
      <Col xs="10">
        <h2>{project.name}</h2>
      </Col>
    );
  }

  return (
    <React.Fragment>
      <Container>
        {card}
        <div className="image-grid">
          <div className="title">
            <Row>
              {header}
              <Col xs="2">
                <PageCount
                  currentPage={currentPage.toString()}
                  totalPages={totalPages.toString()}
                />
              </Col>
            </Row>
          </div>
          <Button
            className="filter-button"
            color="info"
            onClick={e => handleShowFilters(e)}
          >
            <FontAwesomeIcon icon={faFilter} />
          </Button>
          {filterComponent}
          <Row>{imagesGrid}</Row>
          <Row>
            <Col className="text-center">{paginationBar}</Col>
          </Row>
        </div>
      </Container>
    </React.Fragment>
  );
};

export default ImageGrid;
