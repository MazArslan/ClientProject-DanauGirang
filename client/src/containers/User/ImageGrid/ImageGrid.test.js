import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";
import ImageGrid from "./ImageGrid";

Enzyme.configure({ adapter: new Adapter() });

// Mocks the useParams hook
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    projectId: "5e63b45c1c9d4400001bb204"
  })
}));

describe("<ImageGrid/>", () => {
  describe("component rendering", () => {
    const container = shallow(<ImageGrid />);
    it("should render correctly", () => {
      expect(container.exists()).toBe(true);
      expect(toJson(container)).toMatchSnapshot();
    });

    it("should render a show filters button", () => {
      expect(
        container
          .find("Button")
          .at(0)
          .props().className
      ).toEqual("filter-button");
      expect(container.find("FontAwesomeIcon").props().icon.iconName).toEqual(
        "filter"
      );
    });

    it("should render pagination bar navigation buttons", () => {
      expect(
        container
          .find("PaginationItem")
          .at(0)
          .props().className
      ).toEqual("navigate-back");
      expect(
        container
          .find("PaginationItem")
          .at(1)
          .props().className
      ).toEqual("navigate-forward");
    });
  });
});
