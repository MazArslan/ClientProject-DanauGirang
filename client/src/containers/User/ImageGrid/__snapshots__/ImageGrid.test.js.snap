// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageGrid/> component rendering should render correctly 1`] = `
<Fragment>
  <Container
    tag="div"
  >
    <div
      className="image-grid"
    >
      <div
        className="title"
      >
        <Row
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <Col
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
            xs="2"
          >
            <pageCount
              currentPage="1"
              totalPages="0"
            />
          </Col>
        </Row>
      </div>
      <Button
        className="filter-button"
        color="info"
        onClick={[Function]}
        tag="button"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f0b0",
                "M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z",
              ],
              "iconName": "filter",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </Button>
      <Row
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      />
      <Row
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Col
          className="text-center"
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <div
            className="paginationWrapper"
          >
            <Pagination
              aria-label="Page navigation"
              listTag="ul"
              tag="nav"
            >
              <PaginationItem
                className="navigate-back"
                disabled={true}
                tag="li"
              >
                <PaginationLink
                  href="#"
                  onClick={[Function]}
                  previous={true}
                  tag="a"
                />
              </PaginationItem>
              <PaginationItem
                className="navigate-forward"
                disabled={true}
                tag="li"
              >
                <PaginationLink
                  href="#"
                  next={true}
                  onClick={[Function]}
                  tag="a"
                />
              </PaginationItem>
            </Pagination>
          </div>
        </Col>
      </Row>
    </div>
  </Container>
</Fragment>
`;
